org.osgi.test.common/src/main/java/org/osgi/test/common/install/InstallBundle.java

```diff
@@ -18,21 +18,294 @@
+		 *
+		 * @param bundleSymbolicName - symbolicName of the bundle that embedded
+		 *            the file must not be null
+		 * @param bundleVersion - version of the bundle that embedded the file -
+		 *            must not be null
+		 * @param path - the path in the jar - must not be null
+		 * @param file - the filename must not be null
+		 * @return EmbeddedLocation
+		 */
+		public static EmbeddedLocation of(String bundleSymbolicName, Version bundleVersion, String path, String file) {
+			return new EmbeddedLocation(bundleSymbolicName, bundleVersion, path, file);
+		}
+
+		private String	bundleSymbolicName;
+
+		private Version	bundleVersion;
+
+		private String	file;
+		private String	path;
+
+		private EmbeddedLocation() {
+			// no access here.
+		}
+
+		private EmbeddedLocation(String bundleSymbolicName, Version bundleVersion, String path, String file) {
+			if (bundleSymbolicName == null) {
+				throw new IllegalArgumentException("bundleSymbolicName must not be null");
+			}
+			if (bundleVersion == null) {
+				throw new IllegalArgumentException("bundleVersion must not be null");
+			}
+			if (path == null) {
+				throw new IllegalArgumentException("path must not be null");
+			}
+			if (file == null) {
+				throw new IllegalArgumentException("file must not be null");
+			}
+			this.bundleSymbolicName = bundleSymbolicName;
+			this.bundleVersion = bundleVersion;
+
+			path = path.startsWith("/") ? path.substring(1) : path;
+			path = path.endsWith("/") ? path.substring(0, path.length() - 1) : path;
+			this.path = path;
+
+			if (file.startsWith("/")) {
+				file = file.substring(1);
+			}
+			this.file = file;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (this == obj)
+				return true;
+			if (obj == null)
+				return false;
+			if (getClass() != obj.getClass())
+				return false;
+			EmbeddedLocation other = (EmbeddedLocation) obj;
+			if (bundleSymbolicName == null) {
+				if (other.bundleSymbolicName != null)
+					return false;
+			} else if (!bundleSymbolicName.equals(other.bundleSymbolicName))
+				return false;
+			if (bundleVersion == null) {
+				if (other.bundleVersion != null)
+					return false;
+			} else if (!bundleVersion.equals(other.bundleVersion))
+				return false;
+			if (file == null) {
+				if (other.file != null)
+					return false;
+			} else if (!file.equals(other.file))
+				return false;
+			if (path == null) {
+				if (other.path != null)
+					return false;
+			} else if (!path.equals(other.path))
+				return false;
+			return true;
+		}
+
+		@Override
+		public int hashCode() {
+			final int prime = 31;
+			int result = 1;
+			result = prime * result + ((bundleSymbolicName == null) ? 0 : bundleSymbolicName.hashCode());
+			result = prime * result + ((bundleVersion == null) ? 0 : bundleVersion.hashCode());
+			result = prime * result + ((file == null) ? 0 : file.hashCode());
+			result = prime * result + ((path == null) ? 0 : path.hashCode());
+			return result;
```

Comment:
You can just use `Objects.hash(bundleSymbolicName, bundleSymbolicName, file, file);`

Reply:
now use Objects methods